BackBone Refactor for LinkInspector

Domily11 opened the pull request:
Changed all the cell references into more appropriate methods of taking in attributes from models and added parameters to the link model so that actor and intention link can use the same model
Issues #445 and #446 outstanding

Domily11 committed:
First Step of BackBone Refactor for LinkInspector

Domily11 committed:
Change the setSelectValue to get attributes directly from model

Also, used it in initialize

Domily11 committed:
Merge branch 'backbone-refactor' of https://github.com/amgrubb/BloomingLeaf into mack-LinkInspectorBackBone

Domily11 committed:
Able to Call LinkInspector again

Manarnz committed:
changed code in render to single quotation marks

Manarnz committed:
Merge branch 'mack-LinkInspectorBackBone' of https://github.com/amgrubb/BloomingLeaf into mack-LinkInspectorBackBone

Domily11 committed:
Fixed problem where weren't able to choose evolving relationships

Domily11 committed:
Merge branch 'mack-LinkInspectorBackBone' of https://github.com/amgrubb/BloomingLeaf into mack-LinkInspectorBackBone

Manarnz committed:
Added the necessary code for clearInspector

Manarnz committed:
Merge branch 'backbone-refactor' of https://github.com/amgrubb/BloomingLeaf into mack-LinkInspectorBackBone

Manarnz committed:
Attempt to switch to switchMode

Manarnz committed:
Fixed bugs from when I switched to switchMode

Domily11 committed:
Fixed dropdown menu problem for evolving relations

The "end" is not greyed out however

Domily11 committed:
Half Fixed the Dropdown Bug

Previously when we clicked switched between evolving and constant relationships the "begin" in the evolving relationship will be unclickable when you switch back to it the second time. Now it is but only if you did click update or choose anything at first in the begin dropdown

Domily11 committed:
Commit to save work

Domily11 committed:
Revert "Commit to save work"

This reverts commit 97c9bf9c4baee5e8ec78bd3bfe34bac678fbfc9b.

Domily11 committed:
Started scripting the template for evolving functions

Domily11 committed:
Fixed template and bug of dropdown menu

Domily11 committed:
Mack testtemplate (#429)

* Create BackboneModels.js

Create file for backbone model conversion

* Temporarily removing Load Sample Model button until backbone refactor is finished (#410)

* Add BackboneModels.js to index.html

* Moved onclick event to onFunctions.js (#412)

* Moved onclick event to onFunctions.js, copy paste functions work

* Moved keyboard functions and deleted old file

* removed keyboardFunctions.js from index.html

* Cleanup displayAnalysis (#422)

* Added backbone model representations for UserEvaluation and Constraint

* Deleted commented lines

* Removed dead functions and commented out lines

* Deleted more functions

* Removed extra file

* Deleted refreshAnalysisUI() switchResults() and switchSelectedShadingResult(). Made issue for refreshAnalysisUI

* Added some TODOs

Co-authored-by: leshghi <43756305+leshghi@users.noreply.github.com>

* Created FuncSegment and RepFuncSegment backbone models in BackboneModels.js (#421)

* made FuncSegment and RepFuncSegment backbone models

* Changed parameters in initialize function to single option parameter

* Updated variable names

* Update RepFuncSegment model name

* Update BackboneModels.js

Co-authored-by: angelac2023 <achu91@smith.edu>
Co-authored-by: A. M. Grubb <amgrubb@smith.edu>

* Added backbone model representations for UserEvaluation and Constraint (#420)

* Added backbone model representations for UserEvaluation and Constraint

* Deleted commented lines

* Made name changes and defaults

* added this.type

* Changed destRefTP name.

* Update BackboneModels.js

* removed defaults

Co-authored-by: A. M. Grubb <amgrubb@smith.edu>

* Tried another way of going with the template

* Commented thought process

* Updated more comments and took out unnecessary codes

* Added comment

Co-authored-by: leshghi <43756305+leshghi@users.noreply.github.com>
Co-authored-by: YesugenB <59470488+YesugenB@users.noreply.github.com>
Co-authored-by: i-m-foster <72589318+i-m-foster@users.noreply.github.com>
Co-authored-by: angelac2023 <achu91@smith.edu>
Co-authored-by: A. M. Grubb <amgrubb@smith.edu>

Domily11 committed:
Deleted extra lines

Domily11 committed:
Merge branch 'backbone-refactor' of https://github.com/amgrubb/BloomingLeaf into mack-LinkInspectorBackBone

Domily11 committed:
Merge branch 'backbone-refactor' of https://github.com/amgrubb/BloomingLeaf into mack-LinkInspectorBackBone

Manarnz committed:
Combining LinkBBM with LinkInspector (#432)

* Getting started

This is just us testing, nothing serious.

* Fixed issues in previous commit

We changed the placement of selected from LinkBBM to the model, as it is concerned with the interface and not backbone.
** In each function, we defiende link, is that okay?

* Fixed linkType and type problem

Another bug came up tho, that when you are switching between different link types it reassignes all the "linkType" in all the links

* Took out actorInspector.clear()

Co-authored-by: Domily11 <47505178+Domily11@users.noreply.github.com>

leshghi commented on leaf-analysis/app.js:
```diff
@@ -4,7 +4,7 @@
 
 // Name of .jar file for BloomingLeaf project must be Blooming.jar
 //var userPath = "/Users/<your user path here>/BloomingLeaf"
-var userPath = "/Users/judySmith/git/BloomingLeaf"
```
Remove this file from commit

leshghi commented on leaf-ui/index.html:
```diff
@@ -71,7 +71,6 @@
                 <a id="legend">Legend</a>
                 <a id="evo-color-key">EVO Color Key</a>
                 <!-- <a id="load-sample">Load Sample Model</a> --> 
-
```
Remove this file from commit

leshghi commented on leaf-ui/js/initializeElements.js:
```diff
@@ -79,11 +79,12 @@ var analysisRequest = new AnalysisRequest();        //TODO: make not global!
 var configCollection = new ConfigCollection([]);
 var configInspector = new ConfigInspector({collection:configCollection});
 
+    
```
Remove line

leshghi commented on leaf-ui/js/initializeElements.js:
```diff
@@ -79,11 +79,12 @@ var analysisRequest = new AnalysisRequest();        //TODO: make not global!
 
 
-var linkInspector = new LinkInspector();
+
```
Remove line

leshghi commented on leaf-ui/js/initializeElements.js:
```diff
@@ -303,8 +304,8 @@ var act = new joint.shapes.basic.Actor({ position: {x: 40, y: 355} });
 
-// Setup LinkInspector
-$('.inspector').append(linkInspector.el);
+
```
Remove lines

leshghi commented on leaf-ui/js/loadSaveFunctions.js:
```diff
@@ -126,6 +126,7 @@ function getLinksArr(arr) {
 	for (var i = 0; i < arr.length; i++) {
 		var link = new Link(arr[i].linkType, arr[i].linkSrcID, arr[i].absoluteValue);
 		link.linkID = arr[i].linkID;
+		console.log("WE ARE DEAD");
```
Ok remove this lmao

leshghi commented on leaf-ui/js/object/modelObjects.js:
```diff
@@ -330,7 +330,8 @@ class Link {
     constructor(linkType, linkSrcID, absoluteValue) {
-        this.linkID = this.createID();
+        //this.linkID = this.createID();
+        console.log("WTF");
```
:( this too

leshghi commented on leaf-ui/rappid-extensions/BackboneModels.js:
```diff
@@ -0,0 +1,66 @@
+/** This file contains backbone model representations of the original model objects - WIP */
```
Pull from backbone-refactor because this file should no longer exist

leshghi commented on leaf-ui/rappid-extensions/LinkInspector.js:
```diff
@@ -1,12 +1,17 @@
 
-    className: 'link-inspec',
+    initialize:function(){
+        this.listenTo(this.model,'change:selected', this.rerender);
```
What does the selected param represent? do we need this?

Domily11 committed:
Merge branch 'backbone-refactor' of https://github.com/amgrubb/BloomingLeaf into mack-LinkInspectorBackBone

Manarnz committed:
Merge branch 'backbone-refactor' of https://github.com/amgrubb/BloomingLeaf into mack-LinkInspectorBackBone

Manarnz committed:
Merge branch 'mack-LinkInspectorBackBone' of https://github.com/amgrubb/BloomingLeaf into mack-LinkInspectorBackBone

leshghi commented on leaf-ui/js/onFunctions.js:
```diff
@@ -733,7 +733,8 @@ $('#colorblind-mode-isOn').on('click', function(){ //turns off colorblind mode
  */
 function createLink(cell) {
```
Remove function

leshghi commented on leaf-ui/js/onFunctions.js:
```diff
@@ -1143,8 +1143,8 @@ graph.on('remove', function(cell) {
 	elementInspector.clear();
-	linkInspector.clear();
-
+	//linkInspector.clear();
```
Remove both these lines

leshghi commented on leaf-ui/rappid-extensions/BackboneModelsOther.js:
```diff
@@ -27,10 +27,11 @@
 
     defaults: {
-        linkDestID: null,
+        //linkDestID: null,
```
Remove this line

leshghi commented on leaf-ui/rappid-extensions/LinkInspector.js:
```diff
@@ -1,12 +1,17 @@
+    },
 
     template: [
+        '<script type="text/template" id="item-template">',
```
You can remove the script tag if you are not using it

leshghi commented on leaf-ui/rappid-extensions/LinkInspector.js:
```diff
@@ -28,103 +33,172 @@ var LinkInspector = Backbone.View.extend({
     ].join(''),
 
     actortemplate: [
+        '<script type="text/template" id="item-template">',
```
Can also remove script tag here if not using it, although this is small enough that maybe it would be easy to use

leshghi replied:
^

leshghi commented on leaf-ui/rappid-extensions/LinkInspector.js:
```diff
@@ -208,151 +281,62 @@ var LinkInspector = Backbone.View.extend({
         return node.prop('.funcvalue/text') == 'NB';
```
This function needs to be re-written as it is throwing an error

leshghi replied:
^

leshghi commented on leaf-ui/rappid-extensions/LinkInspector.js:
```diff
@@ -28,103 +33,172 @@ var LinkInspector = Backbone.View.extend({
+                link.set('evolving', false); //makes sure the rerender doesn't activate
+                this.$el.html(_.template($(this.template).html())(this.model.toJSON()));
+                $('#constant-links').val(link.get('linkType'));
+                //didn't put updateConstantRelationship here because if it was rendered everything will be the same as previous and if previous was already constant it would not need new update until begin is changed       
```
Can delete

leshghi commented on leaf-ui/rappid-extensions/LinkInspector.js:
```diff
@@ -28,103 +33,172 @@ var LinkInspector = Backbone.View.extend({
-        this.cell = cell;
-        this.link = model.getLinkByID(cell.attributes.linkID);
+    render: function() {
+        console.log(this.model)
```
Remove this

leshghi commented on leaf-ui/rappid-extensions/LinkInspector.js:
```diff
@@ -28,103 +33,172 @@ var LinkInspector = Backbone.View.extend({
-    render: function(cell) {
-        this.cell = cell;
-        this.link = model.getLinkByID(cell.attributes.linkID);
+    render: function() {
```
Look into any way to simplify this render function to make it cleaner/more readable.

leshghi replied:
^

leshghi commented on leaf-ui/rappid-extensions/LinkInspector.js:
```diff
@@ -28,103 +33,172 @@ var LinkInspector = Backbone.View.extend({
     },
 
+    /**
+     * rerender to update the html so that appropriate options are hidden/shown or disabled for evolving relations
```
```suggestion
     * Rerender function to update the html template so that appropriate options are allowed for evolving relations
```

leshghi replied:
^

leshghi commented on leaf-ui/rappid-extensions/LinkInspector.js:
```diff
@@ -28,103 +33,172 @@ var LinkInspector = Backbone.View.extend({
+     * rerender to update the html so that appropriate options are hidden/shown or disabled for evolving relations
+     * Only for evolving template as that is the only one that needs constant update for options
+     * 
+     * This function is called on when selected is changed
```
Does selected represent if the evolving template should be used? If so, it should be removed and you should use the "evolving" param in LinkBBM instead.

leshghi commented on leaf-ui/rappid-extensions/LinkInspector.js:
```diff
@@ -28,103 +33,172 @@ var LinkInspector = Backbone.View.extend({
-        this.link = model.getLinkByID(cell.attributes.linkID);
+    render: function() {
+        console.log(this.model)
+        var link = this.model.get('link');
```
Instead of calling this.model.get('link') everywhere, write an initialize function that sets this.link = this.model.get('link')

leshghi replied:
^

leshghi commented on leaf-ui/rappid-extensions/LinkInspector.js:
```diff
@@ -28,103 +33,172 @@ var LinkInspector = Backbone.View.extend({
+    },
+
+    /**
+     * remove the previous view so that we don't have multiple LinkInspector showing up
```
```suggestion
     * Remove the previous view so that only one LinkInspector View is present in the 
     *  DOM at a time
```

leshghi replied:
^

leshghi commented on leaf-ui/rappid-extensions/LinkInspector.js:
```diff
@@ -28,103 +33,172 @@ var LinkInspector = Backbone.View.extend({
+    /**
+     * remove the previous view so that we don't have multiple LinkInspector showing up
+     *
+     * This function is called when clearInspector is called
```
```suggestion
```

leshghi replied:
^

leshghi commented on leaf-ui/rappid-extensions/LinkInspector.js:
```diff
@@ -28,103 +33,172 @@ var LinkInspector = Backbone.View.extend({
+        var link = this.model.get('link');
+        var type = link.get('linkType');
+        var postType = link.get('postType');
+        var current = link.get('evolving');
```
Don't need a variable for something you only access once
```suggestion
```

leshghi replied:
^

leshghi commented on leaf-ui/rappid-extensions/LinkInspector.js:
```diff
@@ -28,103 +33,172 @@ var LinkInspector = Backbone.View.extend({
      *
      * This function is called on click for #switch-link-type
      */
     switchMode: function() {
```
Logic in this function can be simplified/cleaned up

leshghi commented on leaf-ui/rappid-extensions/LinkInspector.js:
```diff
@@ -28,103 +33,172 @@ var LinkInspector = Backbone.View.extend({
+        var current = link.get('evolving');
+        link.set('evolving', !current);
+        if(link.get('evolving')){
+            this.$el.html(_.template($(this.evolvingtemplate).html())(this.model.toJSON()));
```
I would suggest using a render function for updating the html template

leshghi commented on leaf-ui/rappid-extensions/LinkInspector.js:
```diff
@@ -28,103 +33,172 @@ var LinkInspector = Backbone.View.extend({
-            $('#constant-links').val(values[0].trim());
-            this.evolvingRelations = false; 
+        }else {
+            link.set('evolving', false);
```
I'm confused about why you changed the value of the evolving param, checked it, and then changed it again

Manarnz replied:
It is not important, we just took it out.

leshghi commented on leaf-ui/rappid-extensions/LinkInspector.js:
```diff
@@ -208,151 +281,62 @@ var LinkInspector = Backbone.View.extend({
-            });
+        var end = link.get('postType');
+        link.set("linkType", begin); //modify the model
+        var option = "#link-type-end option[value= \"" + begin + "\"]"; //for hiding the already chosen value at the beginning later on in line 294
```
Why are we hiding the already chosen value? Also - no need to put line numbers in comments

Manarnz replied:
It is not important, we just took it out.

Manarnz replied:
Previously it was like that and we did not want the begin and end to be the same, this would defeat the purpose of evolving relationship.

leshghi replied:
Since this is only used once I would put it directly in the $(option).hide(); line (315) and comment // Hides the selected begin value

leshghi commented on leaf-ui/rappid-extensions/LinkInspector.js:
```diff
@@ -208,151 +281,62 @@ var LinkInspector = Backbone.View.extend({
+        
+        //makes the words on the links into lower case
+        if (['AND', 'OR', 'NO'].includes(begin)){
+            begin = begin.toLowerCase();
```
You can use toLowerCase() on any string even if it has characters, so it might be simpler to just directly convert to lower case

leshghi commented on leaf-ui/rappid-extensions/LinkInspector.js:
```diff
@@ -208,151 +281,62 @@ var LinkInspector = Backbone.View.extend({
-            element.html('<option class="select-placeholder" selected disabled value="">End</option>');
+        link.set("postType", end);
+        if (['AND', 'OR', 'NO'].includes(begin)){
+            begin = begin.toLowerCase();
```
Same here re: toLowerCase

leshghi reviewed (changes requested):
Great work! See comments :)

Manarnz committed:
Fixed the actorType issue and aded the error message

Previously, when the target is initally connected, it is not updated. To solve this, we created the updateTarget. We also added a feature, such that if the source is actor and the target is not, an error message would show up instead of the actor template.

Manarnz committed:
implemented Ali's comments

Manarnz committed:
deleted the app.js file

Manarnz committed:
deleted index.html

leshghi commented on leaf-ui/js/analysis.js:
```diff
@@ -51,14 +51,14 @@ function init(){
     graph = jQuery.extend({}, window.opener.graph);
     var i = analysisRequest.currentState.indexOf('|', 0);
     current = parseInt(analysisRequest.currentState.substring(0, i));
-
```
Remove this file. This is for the next states feature which we will be touching/updating last

leshghi commented on leaf-ui/js/onFunctions.js:
```diff
@@ -942,47 +908,14 @@ paper.on("link:options", function(cell, evt){
 
 	linkInspector.render();
 
+    
```
Remove this line

leshghi committed:
Remove unwanted files from PR, fix smaller comments that were overlooked

leshghi commented on leaf-ui/rappid-extensions/LinkInspector.js:
```diff
@@ -208,151 +286,55 @@ var LinkInspector = Backbone.View.extend({
+        
+        //makes the words on the links into lower case
+        begin = begin.toLowerCase();
+        end = end.toLowerCase();
```
Has end been set here? Shouldn't it be null most of the time at this point?

leshghi committed:
Merge branch 'backbone-refactor' of https://github.com/amgrubb/BloomingLeaf into mack-LinkInspectorBackBone

leshghi commented on leaf-ui/rappid-extensions/LinkInspector.js:
```diff
@@ -28,103 +33,177 @@ var LinkInspector = Backbone.View.extend({
-            '<select id="link-type-end" class="repeat-select">',
+            '<select id="link-type-end" class="repeat-select" <% if (!selected) {%> disabled= true style = "background-color:gray"<%} %>>',
+                '<option value="NO">No Relationship</option>',
+                '<option value="AND" class = "A">And-Decomposition</option>', //class works with function in line 294 and 296
```
Remove line numbers

leshghi commented on leaf-ui/rappid-extensions/LinkInspector.js:
```diff
@@ -28,103 +33,177 @@ var LinkInspector = Backbone.View.extend({
 
     events: {
-            'click #switch-link-type': 'switchMode',
+            'click #switch-link-type': 'renderNewMode',
```
What if you put a separate id for the switch button in the evolving template vs the constant template here instead? That way each id could call a different function that either switched to constant mode or to evolving mode depending. This would simplify a lot of your logic

leshghi commented on leaf-ui/rappid-extensions/LinkInspector.js:
```diff
@@ -28,103 +33,177 @@ var LinkInspector = Backbone.View.extend({
+        link.set('evolving', !current);
+        if(link.get('evolving')){
+            this.$el.html(_.template($(this.evolvingtemplate).html())(this.model.toJSON()));
+            //if the postType is already chosen, meaning not the first time visiting LinkInspector, take the previous values
```
If you switch from constant to evolving, postType should always be null

leshghi commented on leaf-ui/rappid-extensions/LinkInspector.js:
```diff
@@ -208,151 +286,55 @@ var LinkInspector = Backbone.View.extend({
+
+        //set the options
+        $('option').show(); //clear the previous selection
+        if(begin == "no"){ //need to specify case for "no" otherwise will be characterized as the else condition
```
This logic can be simpler, just remove the if(begin==no) then have if and else if statements for the rest of the logic

leshghi reviewed (changes requested):
Once you respond to a comment and commit that, please resolve the comment so I know you have dealt with the issue! Also, please standardize the quotation marks in this file - you should be using single quotes ('string here') when possible. I started checking out a lot of these issues on a separate branch and made some progress so if you need help/want to check out what I've done there let me know

leshghi committed:
Some clean up for link inspector (#444)

* first round of cleanup

* More cleanup to LinkInspector.js

* Added structural changes for implementation of intentionBBM (#440)

* Reorg and comments

* Fix small bugs

* Remove repeat error since it does not accurately potray when link data is saved

Co-authored-by: i-m-foster <72589318+i-m-foster@users.noreply.github.com>

leshghi committed:
Merge branch 'backbone-refactor' of https://github.com/amgrubb/BloomingLeaf into mack-LinkInspectorBackBone

leshghi committed:
Last fixes post merge
